pandas/core/generic.py

```diff
@@ -2930,25 +2931,61 @@ def blocks(self):
         """Internal property, property synonym for as_blocks()"""
         return self.as_blocks()
 
-    def astype(self, dtype, copy=True, raise_on_error=True, **kwargs):
+    def astype(self, dtype, copy=True, inplace=False, raise_on_error=True,
+               **kwargs):
         """
         Cast object to input numpy.dtype
-        Return a copy when copy = True (be really careful with this!)
 
         Parameters
         ----------
-        dtype : numpy.dtype or Python type
+        dtype : numpy.dtype or Python type (to cast entire DataFrame to the
+            same type). Alternatively, {col: dtype, ...}, where col is a column
+            label and dtype is a numpy.dtype or Python type (to cast one or
+            more of the DataFrame's columns to column-specific types).
+        copy : deprecated; use inplace instead
+        inplace : boolean, default False
+            Modify the NDFrame in place (do not create a new object)
         raise_on_error : raise on invalid input
-        kwargs : keyword arguments to pass on to the constructor
+        kwargs : keyword arguments to pass on to the constructor if
+            inplace=False
 
         Returns
         -------
-        casted : type of caller
-        """
-
-        mgr = self._data.astype(dtype=dtype, copy=copy,
-                                raise_on_error=raise_on_error, **kwargs)
-        return self._constructor(mgr).__finalize__(self)
+        casted : type of caller (if inplace=False) or None (if inplace=True)
+        """
+        if isinstance(dtype, collections.Mapping):
+            if self.ndim == 1:  # i.e. Series
+                if len(dtype) > 1 or list(dtype.keys())[0] != self.name:
+                    if raise_on_error:
+                        raise KeyError('Only the Series name can be used for '
+                                       'the key in Series dtype mappings.')
+                    return
+                for key, value in dtype.items():
+                    return self.astype(value, copy, inplace, raise_on_error,
+                                       **kwargs)
+
+            if inplace:
+                for col, typ in dtype.items():
+                    self[col].astype(typ, inplace=True,
+                                     raise_on_error=raise_on_error)
+                return
+            from pandas.tools.merge import concat
+            casted_cols = [self[col].astype(typ, copy=copy)
+                           for col, typ in dtype.items()]
+            other_col_labels = self.columns.difference(dtype.keys())
```

Comment:
do the astype on each column, NOT inplace. Then assemble them. Then if its inplace update, else return it.